File: archestra-bench/envs.py
"""Load benchmark environments from TOML.

An environment (`envs/<id>.toml`) bundles a single agent, a web-pinned skill surface, remote MCP
servers, and the ids of the tasks that run against it. Tasks themselves are self-contained under
`tasks/<id>/` (see tasks.py); the env only references them by id.

Validation is loud: any malformed or missing field raises SystemExit naming the offending file/task,
so a misconfigured environment never degrades into a silently partial run.
"""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import tomlconf
from tasks import Task, load_task

_DEFAULT_SYSTEM_PROMPT = "You are an expert software engineer completing a benchmark task."
_TOOL_SHORT_RE = re.compile(r"[a-z][a-z0-9_]*")  # archestra built-in short name (e.g. create_skill)


@dataclass(frozen=True)
class SkillRef:
    """a web skill ref imported pinned to `ref` (commit/branch/tag). `path` scopes discovery.

    `cap` bounds how many discovered skills are imported; None imports all of them."""

    repo: str
    path: str | None
    ref: str
    cap: int | None = None


@dataclass(frozen=True)
class Mcp:
    """a remote (HTTP) MCP server the agent may use, seeded as a remote catalog item by URL."""

    name: str
    server_url: str


@dataclass(frozen=True)
class EnvConfig:
    """one environment: a single agent, its skill/mcp surface, and the tasks that run in it."""

    id: str
    name: str
    agent_name: str
    agent_system_prompt: str
    skills: tuple[SkillRef, ...]
    mcps: tuple[Mcp, ...]
    tasks: tuple[Task, ...]
    tools: tuple[str, ...] = ()  # extra archestra__* short names to assign (e.g. create_skill)
    # when true, all lanes of this env share one backend (seeded once); each lane still gets its own
    # agent + benchmark MCP. Only safe for envs whose tasks do NOT mutate shared backend state -- a
    # mutating env (e.g. one that creates/counts skills) must stay isolated (default) so concurrent
    # lanes never desync. See run.py's lane execution.
    share_backend: bool = False


def load_envs(envs_dir: Path) -> dict[str, EnvConfig]:
    """Load every `envs/*.toml`, validating env ids are db-slug-safe and task ids globally unique.

    Task ids referenced by an env resolve to `tasks/<id>/` under the bench root (the directory
    containing `envs/`)."""
    root = envs_dir.parent
    files = sorted(envs_dir.glob("*.toml"))
    if not files:
        raise SystemExit(f"no environment files found in {envs_dir}")
    envs: dict[str, EnvConfig] = {}
    task_owner: dict[str, str] = {}
    for path in files:
        env = _load_env(path, root)
        if env.id in envs:
            raise SystemExit(f"duplicate environment id {env.id!r} (in {path.name})")
        for task in env.tasks:
            if task.id in task_owner:
                raise SystemExit(
                    f"task id {task.id!r} is defined in both {task_owner[task.id]!r} and {env.id!r}; "
                    "task ids must be globally unique across environments"
                )
            task_owner[task.id] = env.id
        envs[env.id] = env
    return envs


def _load_env(path: Path, root: Path) -> EnvConfig:
    ctx = path.name
    data = tomlconf.parse_toml(path)
    env_id = tomlconf.req_str(data, "id", ctx)
    if not tomlconf.is_slug(env_id):
        raise SystemExit(f"{ctx}: env id {env_id!r} must be lowercase alphanumeric with dashes (db-slug-safe)")
    name = tomlconf.req_str(data, "name", ctx, default=env_id)
    agent = tomlconf.table(data, "agent", ctx, default={})
    agent_name = tomlconf.req_str(agent, "name", f"{ctx} [agent]", default=f"{env_id}-agent")
    agent_prompt = tomlconf.req_str(agent, "system_prompt", f"{ctx} [agent]", default=_DEFAULT_SYSTEM_PROMPT)
    skills = tuple(_skill_ref(row, f"{ctx} [[skills]]") for row in tomlconf.rows(data, "skills", ctx))
    mcps = _mcps(tomlconf.rows(data, "mcps", ctx), f"{ctx} [[mcps]]")
    task_ids = tomlconf.strs(data, "tasks", ctx)
    if not task_ids:
        raise SystemExit(f"{ctx}: environment {env_id!r} declares no tasks")
    for task_id in task_ids:
        if not tomlconf.is_slug(task_id):
            raise SystemExit(f"{ctx}: task id {task_id!r} must be lowercase alphanumeric with dashes (slug-safe)")
    tasks = tuple(load_task((root / "tasks" / task_id).resolve()) for task_id in task_ids)
    return EnvConfig(
        id=env_id,
        name=name,
        agent_name=agent_name,
        agent_system_prompt=agent_prompt,
        skills=skills,
        mcps=mcps,
        tasks=tasks,
        tools=_tool_names(tomlconf.strs(data, "tools", ctx), f"{ctx} tools"),
        share_backend=tomlconf.opt_bool(data, "share_backend", ctx),
    )


def _tool_names(names: list[str], ctx: str) -> tuple[str, ...]:
    """validate each extra tool is an archestra built-in short name; reject duplicates."""
    for name in names:
        if not _TOOL_SHORT_RE.fullmatch(name):
            raise SystemExit(f"{ctx}: tool {name!r} must be a lowercase archestra short name (e.g. create_skill)")
    if len(names) != len(set(names)):
        raise SystemExit(f"{ctx}: duplicate tool name(s) in {names}")
    return tuple(names)


def _skill_ref(row: Mapping[str, Any], ctx: str) -> SkillRef:
    cap = tomlconf.opt_int(row, "cap", ctx)
    if cap is not None and cap < 1:
        raise SystemExit(f"{ctx}: cap must be >= 1, got {cap}")
    ref = tomlconf.req_str(row, "ref", ctx)
    # the pin is carried as `.../tree/<ref>`, which cannot represent a ref containing a slash
    # (e.g. a `feature/x` branch) -- use a commit SHA or a slash-free tag.
    if "/" in ref:
        raise SystemExit(f"{ctx}: ref {ref!r} must not contain '/' (use a commit SHA or a slash-free tag)")
    return SkillRef(repo=tomlconf.req_str(row, "repo", ctx), path=tomlconf.opt_str(row, "path", ctx), ref=ref, cap=cap)


def _mcps(mcp_rows: list[Mapping[str, Any]], ctx: str) -> tuple[Mcp, ...]:
    mcps = tuple(
        Mcp(name=tomlconf.req_str(r, "name", ctx), server_url=tomlconf.req_str(r, "server_url", ctx))
        for r in mcp_rows
    )
    names = [m.name for m in mcps]
    if len(names) != len(set(names)):
        raise SystemExit(f"{ctx}: duplicate MCP name(s) in {names}")
    return mcps

File: archestra-bench/runner/src/config/envs.rs
use std::collections::HashMap;
use std::path::Path;
use std::str::FromStr;

use super::tasks::TaskConfigError;
use super::toml_util::{self, TomlTable};
use super::types::{EnvConfig, Mcp, PlatformConfig, SkillRef, ToolExposureMode};

// Empty by default: lanes mimic a regular Archestra user who sets no custom system prompt, so the
// agent runs on Archestra's stock chat instructions. Submission guidance is appended to the
// final-stage user message instead (see SUBMIT_INSTRUCTION in run.rs).
const DEFAULT_SYSTEM_PROMPT: &str = "";

#[derive(Debug, thiserror::Error)]
#[error("env config error: {0}")]
pub struct EnvConfigError(pub String);

impl From<toml_util::TomlError> for EnvConfigError {
    fn from(e: toml_util::TomlError) -> Self {
        Self(e.to_string())
    }
}

impl From<TaskConfigError> for EnvConfigError {
    fn from(e: TaskConfigError) -> Self {
        Self(e.to_string())
    }
}

pub fn load_envs(envs_dir: &Path) -> Result<HashMap<String, EnvConfig>, EnvConfigError> {
    let root = envs_dir
        .parent()
        .ok_or_else(|| EnvConfigError(format!("envs dir {envs_dir:?} has no parent")))?;

    let mut entries: Vec<_> = std::fs::read_dir(envs_dir)
        .map_err(|e| EnvConfigError(format!("cannot read envs dir: {e}")))?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().is_some_and(|ext| ext == "toml"))
        .collect();
    entries.sort();

    if entries.is_empty() {
        return Err(EnvConfigError(format!(
            "no environment files found in {envs_dir:?}"
        )));
    }

    let mut envs: HashMap<String, EnvConfig> = HashMap::new();
    let mut task_owner: HashMap<String, String> = HashMap::new();

    for path in entries {
        let env = load_env(&path, root)?;
        if let Some(_prev) = envs.insert(env.id.clone(), env.clone()) {
            return Err(EnvConfigError(format!(
                "duplicate environment id {} (in {})",
                env.id,
                path.display()
            )));
        }
        for task in &env.tasks {
            if let Some(prev_env) = task_owner.insert(task.id.clone(), env.id.clone()) {
                return Err(EnvConfigError(format!(
                    "task id {} is defined in both {prev_env:?} and {:?}; task ids must be globally unique across environments",
                    task.id, env.id
                )));
            }
        }
    }

    Ok(envs)
}

fn load_env(path: &Path, root: &Path) -> Result<EnvConfig, EnvConfigError> {
    let ctx = path
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    let data = toml_util::parse_toml_file(path)?;

    let env_id = toml_util::req_str(&data, "id", &ctx)?;
    if !archestra_bench_core::is_slug(&env_id) {
        return Err(EnvConfigError(format!(
            "{ctx}: env id {env_id:?} must be lowercase alphanumeric with dashes (db-slug-safe)"
        )));
    }

    let name = toml_util::req_str_with_default(&data, "name", &ctx, &env_id)?;
    let agent = toml_util::table_with_default(&data, "agent", &ctx)?;
    let agent_ctx = format!("{ctx} [agent]");
    let agent_name =
        toml_util::req_str_with_default(&agent, "name", &agent_ctx, format!("{env_id}-agent"))?;
    let agent_prompt = toml_util::req_str_with_default(
        &agent,
        "system_prompt",
        &agent_ctx,
        DEFAULT_SYSTEM_PROMPT,
    )?;

    let skills = load_skills(&data, &ctx)?;
    let mcps = load_mcps(&data, &ctx)?;

    let task_ids = toml_util::strs(&data, "tasks", &ctx)?;
    if task_ids.is_empty() {
        return Err(EnvConfigError(format!(
            "{ctx}: environment {env_id:?} declares no tasks"
        )));
    }
    for task_id in &task_ids {
        if !archestra_bench_core::is_slug(task_id) {
            return Err(EnvConfigError(format!(
                "{ctx}: task id {task_id:?} must be lowercase alphanumeric with dashes (slug-safe)"
            )));
        }
    }

    let mut tasks = Vec::with_capacity(task_ids.len());
    for task_id in task_ids {
        let task_dir = root
            .join("tasks")
            .join(&task_id)
            .canonicalize()
            .map_err(|e| {
                EnvConfigError(format!(
                    "{ctx}: cannot resolve task dir for {task_id:?}: {e}"
                ))
            })?;
        tasks.push(super::tasks::load_task(&task_dir)?);
    }

    let tools = tool_names(
        &toml_util::strs(&data, "tools", &ctx)?,
        &format!("{ctx} tools"),
    )?;
    let share_backend = toml_util::opt_bool(&data, "share_backend", &ctx, false)?;

    let platform = load_platform(&data, &ctx)?;

    Ok(EnvConfig {
        id: env_id,
        name,
        agent_name,
        agent_system_prompt: agent_prompt,
        skills,
        mcps,
        tasks,
        tools,
        share_backend,
        platform,
    })
}

fn load_platform(data: &TomlTable, ctx: &str) -> Result<PlatformConfig, EnvConfigError> {
    let platform = toml_util::table_with_default(data, "platform", ctx)?;
    let platform_ctx = format!("{ctx} [platform]");
    let mode = toml_util::req_str_with_default(
        &platform,
        "tool_exposure_mode",
        &platform_ctx,
        ToolExposureMode::default().as_str(),
    )?;
    let tool_exposure_mode = ToolExposureMode::from_str(&mode)
        .map_err(|e| EnvConfigError(format!("{platform_ctx}: {e}")))?;
    Ok(PlatformConfig { tool_exposure_mode })
}

fn load_skills(data: &TomlTable, ctx: &str) -> Result<Vec<SkillRef>, EnvConfigError> {
    let rows = toml_util::rows(data, "skills", ctx)?;
    let mut skills = Vec::with_capacity(rows.len());
    for (i, row) in rows.iter().enumerate() {
        let row_ctx = format!("{ctx} [[skills]][{i}]");
        skills.push(load_skill_ref(row, &row_ctx)?);
    }
    Ok(skills)
}

fn load_skill_ref(row: &TomlTable, ctx: &str) -> Result<SkillRef, EnvConfigError> {
    let cap = toml_util::opt_int(row, "cap", ctx)?.map(|c| c as usize);
    if let Some(c) = cap
        && c < 1
    {
        return Err(EnvConfigError(format!("{ctx}: cap must be >= 1, got {c}")));
    }
    let ref_ = toml_util::req_str(row, "ref", ctx)?;
    if ref_.contains('/') {
        return Err(EnvConfigError(format!(
            "{ctx}: ref {ref_:?} must not contain '/' (use a commit SHA or a slash-free tag)"
        )));
    }
    Ok(SkillRef {
        repo: toml_util::req_str(row, "repo", ctx)?,
        path: toml_util::opt_str(row, "path", ctx)?,
        ref_,
        cap,
    })
}

fn load_mcps(data: &TomlTable, ctx: &str) -> Result<Vec<Mcp>, EnvConfigError> {
    let rows = toml_util::rows(data, "mcps", ctx)?;
    let mut mcps = Vec::with_capacity(rows.len());
    let mut names = std::collections::HashSet::new();
    for (i, row) in rows.iter().enumerate() {
        let row_ctx = format!("{ctx} [[mcps]][{i}]");
        let name = toml_util::req_str(row, "name", &row_ctx)?;
        if !names.insert(name.clone()) {
            return Err(EnvConfigError(format!(
                "{row_ctx}: duplicate MCP name {name:?}"
            )));
        }
        mcps.push(Mcp {
            name,
            server_url: toml_util::req_str(row, "server_url", &row_ctx)?,
        });
    }
    Ok(mcps)
}

fn tool_names(names: &[String], ctx: &str) -> Result<Vec<String>, EnvConfigError> {
    for name in names {
        if name.is_empty() {
            return Err(EnvConfigError(format!(
                "{ctx}: tool name must be a lowercase archestra short name (e.g. create_skill)"
            )));
        }
        let mut chars = name.chars();
        if !chars.next().is_some_and(|c| c.is_ascii_lowercase()) {
            return Err(EnvConfigError(format!(
                "{ctx}: tool {name:?} must be a lowercase archestra short name (e.g. create_skill)"
            )));
        }
        if !chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_') {
            return Err(EnvConfigError(format!(
                "{ctx}: tool {name:?} must be a lowercase archestra short name (e.g. create_skill)"
            )));
        }
    }
    let unique: std::collections::HashSet<_> = names.iter().collect();
    if unique.len() != names.len() {
        return Err(EnvConfigError(format!(
            "{ctx}: duplicate tool name(s) in {names:?}"
        )));
    }
    Ok(names.to_vec())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn write_env(
        tmp: &std::path::Path,
        id: &str,
        tasks: &[&str],
        share_backend: bool,
        tools: &[&str],
    ) {
        let mut content = format!(
            r#"id = "{}"
name = "{}"
tasks = {:?}
share_backend = {}
"#,
            id, id, tasks, share_backend
        );
        if !tools.is_empty() {
            content.push_str(&format!("tools = {:?}\n", tools));
        }
        content.push_str("[agent]\nsystem_prompt = \"p\"\n");
        std::fs::write(tmp.join(format!("{id}.toml")), content).unwrap();
    }

    fn make_task_dir(root: &std::path::Path, task_id: &str) {
        let d = root.join("tasks").join(task_id);
        std::fs::create_dir_all(&d).unwrap();
        std::fs::write(
            d.join("task.toml"),
            "[[stages]]\ntext = \"go\"\n[result_schema]\ntype = \"object\"\n",
        )
        .unwrap();
        std::fs::write(d.join("verifier.py"), "def test_ok(): assert True\n").unwrap();
    }

    #[test]
    fn test_load_envs_basic() {
        let tmp = tempfile::tempdir().unwrap();
        let envs_dir = tmp.path().join("envs");
        std::fs::create_dir(&envs_dir).unwrap();
        let tasks_dir = tmp.path().join("tasks");
        std::fs::create_dir(&tasks_dir).unwrap();
        write_env(&envs_dir, "basic", &["t1"], true, &[]);
        write_env(&envs_dir, "api", &["t2"], false, &["create_skill"]);
        make_task_dir(tmp.path(), "t1");
        make_task_dir(tmp.path(), "t2");
        let envs = load_envs(&envs_dir).unwrap();
        assert!(envs.contains_key("basic"));
        assert!(envs.contains_key("api"));
        assert!(envs["basic"].share_backend);
        assert!(!envs["api"].share_backend);
        assert_eq!(envs["api"].tools, vec!["create_skill"]);
    }

    // Write an env that declares task `t1` plus a verbatim `[platform]` fragment, then load it.
    fn load_env_with_platform(platform_fragment: &str) -> Result<EnvConfig, EnvConfigError> {
        let tmp = tempfile::tempdir().unwrap();
        make_task_dir(tmp.path(), "t1");
        let path = tmp.path().join("e.toml");
        let content = format!("id = \"e\"\nname = \"e\"\ntasks = [\"t1\"]\n{platform_fragment}");
        std::fs::write(&path, content).unwrap();
        load_env(&path, tmp.path())
    }

    #[test]
    fn test_platform_defaults_to_search_and_run_only() {
        let env = load_env_with_platform("").unwrap();
        assert_eq!(
            env.platform.tool_exposure_mode,
            ToolExposureMode::SearchAndRunOnly
        );
    }

    #[test]
    fn test_platform_full_is_parsed() {
        let env = load_env_with_platform("[platform]\ntool_exposure_mode = \"full\"\n").unwrap();
        assert_eq!(env.platform.tool_exposure_mode, ToolExposureMode::Full);
    }

    #[test]
    fn test_platform_unknown_mode_errors() {
        let err =
            load_env_with_platform("[platform]\ntool_exposure_mode = \"nope\"\n").unwrap_err();
        assert!(err.0.contains("tool_exposure_mode"), "{}", err.0);
        assert!(err.0.contains("nope"), "{}", err.0);
    }

    #[test]
    fn test_platform_non_string_mode_errors() {
        let err = load_env_with_platform("[platform]\ntool_exposure_mode = 7\n").unwrap_err();
        assert!(err.0.contains("must be a string"), "{}", err.0);
    }

    #[test]
    fn test_platform_non_table_errors() {
        let err = load_env_with_platform("platform = \"x\"\n").unwrap_err();
        assert!(err.0.contains("[platform] must be a table"), "{}", err.0);
    }

    #[test]
    fn test_tool_names_rejects_non_short_name() {
        let err = tool_names(&["Create-Skill".to_string()], "ctx").unwrap_err();
        assert!(err.0.contains("short name"));
    }

    #[test]
    fn test_tool_names_rejects_duplicates() {
        let err = tool_names(
            &["create_skill".to_string(), "create_skill".to_string()],
            "ctx",
        )
        .unwrap_err();
        assert!(err.0.contains("duplicate"));
    }
}
